---
title: "Gene Expression Analysis" 
author: "Author: Your Name"
date: "Last update: `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: true
    toc_float:
        collapsed: true
        smooth_scroll: true
    toc_depth: 3
    fig_caption: yes
    code_folding: show
    number_sections: false
fontsize: 14pt
bibliography: bibtex.bib
type: docs
weight: 304
---

<!---
- Compile from command-line
Rscript -e "rmarkdown::render('exprAnalyisDemo.Rmd', c('html_document'), clean=FALSE)"
-->

# Initialize project 

The following provides brief instructions how to access RStudio Cloud, clone
the pregenerated code for this exercise from GitHub, and run the provided
analysis code. The latter includes the download of genome-wide expression data
from NCBI's GEO database, and then identify differentially expressed genes
(DEGs), enriched GO terms and clusters of co-expressed genes.

1. Create a free personal account on RStudio Cloud [here](https://rstudio.cloud/)
2. Go to RStudio Cloud [here](https://rstudio.cloud/projects/)
3. Select under blue _New Project_ menu: `New Project from Git Repository`
4. Provide this URL for BIOL119 repos: https://github.com/tgirke/BIOL119.git 
5. Optional: to work with your own GitHub repos and push changes to it, 
   it will be necessary to set your GitHub identity with:
   - `git config --global user.email "..."`
   - `git config --global user.name "..."`

# Install required packages
```{r package_installs, eval=FALSE, message=FALSE, warning=FALSE}
if(!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
BiocManager::install(c("GEOquery", "limma", "clusterProfiler", "org.Hs.eg.db", "enrichplot", "pheatmap", "RColorBrewer"))
```

# Import data from GEO

The following uses the `GEOquery` package to automatically download from GEO 
a expression data set that involves a chemical treatment. The specific data set chosen for this exercise is experiment [GDS2778](https://www.ncbi.nlm.nih.gov/sites/GDSbrowser?acc=GDS2778). 
This Affymetrix gene expression profiling experiment from Gillis _et al._ [-@Gillis2007-qk] aims to study the effect of the air pollutant 
1,2,4-benzenetriol on blood cells to provide insight into the molecular basis of benzene cytotoxicity. 
More detailed annotation information about this experiment is available under Reference 
Series [GSE7664](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE7664). 

```{r import_data_from_geo, eval=TRUE, message=FALSE}
library(GEOquery) 
gds <- getGEO("GDS2778") # Download GDS2778 data from GEO
eset <- GDS2eSet(gds)
pData(eset)[ ,1:2] # Inspect experimental design 
exprs(eset)[1:4, ] # Inspect expression data that are RMA normalized and on log2 scale
```

# DEG analysis with Limma

Text and citation... [@Ritchie2015-rl]

```{r deg_analysis, eval=TRUE, message=FALSE}
library(limma) # Loads limma library.
targets <- pData(eset)
design <- model.matrix(~ -1+factor(as.character(targets$agent)))
colnames(design) <- c("Treatment", "Control") 
fit <- lmFit(eset, design) # Fits a linear model for each gene based on the given series of arrays.
contrast.matrix <- makeContrasts(Control-Treatment, levels=design)
fit2 <- contrasts.fit(fit, contrast.matrix) # Computes estimated coefficients and standard errors for a given set of contrasts.
fit2 <- eBayes(fit2)
deg_df <- topTable(fit2, coef=1, adjust="fdr", sort.by="B", number=Inf)
deg_df <- deg_df[!is.na(deg_df$adj.P.Val),]
```

# Enrichment analysis of GO terms 

The following performs over-represention analysis (ORA) of GO terms using functions from 
the `clusterProfiler` package [@Yu2012-hr].

## Prepare input

```{r overrepresentation_analysis, eval=TRUE, message=FALSE}
cutoff <- 0.05 # Cutoff to use for filtering on adjusted p-value (FDR)
ids <- deg_df[deg_df$adj.P.Val <= cutoff, "Gene.ID"]
ids <- ids[!grepl("/", ids)] # Removes ambiguous mappings
ids <- ids[grepl("", ids)]
```

## Enrichment of MF terms

```{r overrepresentation_analysis_mf, eval=TRUE, message=FALSE}
library(clusterProfiler); library(org.Hs.eg.db); library(enrichplot)
ego_mf <- enrichGO(gene=ids, OrgDb=org.Hs.eg.db, ont="MF", pAdjustMethod="BH", pvalueCutoff=0.1, readable=TRUE)
dim(ego_mf)
head(ego_mf)
```

## Visualization of MF result

```{r plot_enrichment_mf, eval=TRUE, message=FALSE}
barplot(ego_mf, showCategory=10)
goplot(ego_mf)
```

## Enrichment of BP terms

```{r overrepresentation_analysis_bp, eval=TRUE, message=FALSE}
ego_bp <- enrichGO(gene=ids, OrgDb=org.Hs.eg.db, ont="BP", pAdjustMethod="BH", pvalueCutoff=0.1, readable=TRUE)
dim(ego_bp)
head(ego_bp)
```

## Visualization of BP result
```{r plot_enrichment_bp, eval=TRUE, message=FALSE}
barplot(ego_bp, showCategory=10)
```

# Clustering

...

```{r clustering, eval=TRUE, message=FALSE}
library(pheatmap); library("RColorBrewer")
cutoff <- 0.05 # Cutoff to use for filtering on adjusted p-value (FDR)
affy_ids <- row.names(deg_df[deg_df$adj.P.Val <= cutoff, ])
deg_ma <- exprs(eset)[affy_ids, ] 
pheatmap(deg_ma, scale="row", color=brewer.pal(9, "Blues"))
```

## Session Info

```{r sessionInfo}
sessionInfo()
```

# References
