---
title: "BIOL119: Gene Expression Analysis" 
author: "Author: Your Name"
date: "Last update: `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: true
    toc_float:
        collapsed: true
        smooth_scroll: true
    toc_depth: 3
    fig_caption: yes
    code_folding: show
    number_sections: true
fontsize: 14pt
bibliography: bibtex.bib
type: docs
weight: 304
---

<!---
- Compile from command-line
Rscript -e "rmarkdown::render('exprAnalyisDemo.Rmd', c('html_document'), clean=FALSE)"
-->

# Get started 

The following provides brief instructions how to access RStudio Cloud, clone
the pregenerated code for this exercise from GitHub, and run the provided
analysis code. The latter downloads genome-wide expression data
from NCBI's GEO database, and then identifies differentially expressed genes
(DEGs), enriched GO terms and clusters of co-expressed genes.

## Initialize project 

To get started with RStudio Cloud, users can follow the itemized instructions
below (steps 1-4). Instead of RStudio Cloud, users can install R from
[CRAN](https://cran.r-project.org) and RStudio from
[here](https://www.rstudio.com/products/rstudio/download/) on their local
computer.  Since the usage of RStudio Cloud does not require any local software
installs, it may be easier for beginners. On the other hand, the local install
option is less restrictive and more flexible.  In both cases the user interfaces are
extremly simmilar and the way to run R code is nearly identical. 

1. Create a free personal account on RStudio Cloud [here](https://rstudio.cloud/)
2. Go to RStudio Cloud [here](https://rstudio.cloud/projects/)
3. Select under blue _New Project_ menu: `New Project from Git Repository`
4. Provide this URL to clone BIOL119 repos: [https://github.com/tgirke/BIOL119.git](https://github.com/tgirke/BIOL119.git) 
5. Optional: to work with your own GitHub repos and push changes to it, 
   it will be necessary to set your GitHub identity with:
   - `git config --global user.email "..."`
   - `git config --global user.name "..."`

## Load R code and run it

The R code along with the human readable text in this document can be opened and executed as follows:

1. Open source `Rmd` file for this document by selecting in top menu: `File ->
   Open File -> exprAnalyisDemo.Rmd` 
2. Execulte code by placing your curser on a line of R code and then pressing
   on your keyboard `Ctrl + Enter`. This will send the corresponding code line
   to the R console and execute (run) the code. It is important to execute the
   lines in the order shown in the text.

Alternatively, one can execute all code in this document at once simply by
pressing the triangle next to the Knit button above the code editor window in RStudio, and
then selecting `Knit to HTML`. This will not only execute the code, but also
regenerate this HTML document and update all output, tables and figures accordingly. 
Similarly, one can generate a PDF document instead. The environment that makes 
this possible is called R Markdown.

# Install required packages

The following will install all packages required for this exercise. A fresh install into 
a new R project will take about 2-3 minutes. This install only needs to be done once, meaning
users want to skip this step after restarting R or rerunning the anlysis.

```{r package_installs, eval=FALSE, message=FALSE, warning=FALSE}
if(!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
BiocManager::install(c("GEOquery", "limma", "clusterProfiler", "org.Hs.eg.db", "enrichplot", "pheatmap", "RColorBrewer"))
```

# Import data from GEO

The following uses the `GEOquery` package to automatically download from GEO 
a expression data set that involves a chemical treatment. The specific data set chosen for this exercise is experiment [GDS2778](https://www.ncbi.nlm.nih.gov/sites/GDSbrowser?acc=GDS2778). 
This Affymetrix gene expression profiling experiment from Gillis _et al._ [-@Gillis2007-qk] aims to study the effect of the air pollutant 
1,2,4-benzenetriol on blood cells to provide insight into the molecular basis of benzene cytotoxicity. 
More detailed annotation information about this experiment is available under Reference 
Series [GSE7664](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE7664). 

```{r import_data_from_geo, eval=TRUE, message=FALSE}
library(GEOquery) 
gds <- getGEO("GDS2778") # Download GDS2778 data from GEO
eset <- GDS2eSet(gds)
```

The experimental design information can be returned with the `pData` function. This includes
sample names, and control, treatment and replicate information.

```{r inspect_data_from_geo_1, eval=TRUE, message=FALSE}
pData(eset)[ ,1:2] 
```
The expression data can be returned with the `exprs` function (here for first 4
rows). In this case the authors have normalized the data already with the RMA
method [@Gautier2004-tk]. This is why the normalization step is skipped here.
Note, RMA returns log2 transformed values which is the expected input format for the
downstream DEG analysis with `limma`. Users who are interested in
learning to perform the normalization themselves want to look into the `rma`
function that is part of the `affy` package from Bioconductor (see
[here](https://bioconductor.org/packages/release/bioc/html/affy.html)).


```{r inspect_data_from_geo_2, eval=TRUE, message=FALSE}
dim(exprs(eset))
exprs(eset)[1:4, ] 
```

# DEG analysis with Limma

The following identifies differentially expressed genes with the `limma`
package [@Ritchie2015-rl]. This includes the following steps: (1) creation of a
design matrix for this data set with `model.matrix`; (2) fitting of a linear
model for each gene based on the given series of arrays with `lmFit`; (3)
creation of a contrast matrix defining the sample comparisons with
`makeContrasts`; (4) estimation of coefficients and standard errors for the
chosen contrast(s) with `contrast.fit`; and (5) computation of moderated
t-statistics and log-odds of differential expression by empirical Bayes
shrinkage of the standard errors towards a common value using `eBayes`.
Subsequently, the final results can be accessed with the `topTable` function.

```{r deg_analysis, eval=TRUE, message=FALSE}
library(limma) # Loads limma library.
targets <- pData(eset)
design <- model.matrix(~ -1+factor(as.character(targets$agent)))
colnames(design) <- c("Treatment", "Control") 
fit <- lmFit(eset, design) 
contrast.matrix <- makeContrasts(Control-Treatment, levels=design)
fit2 <- contrasts.fit(fit, contrast.matrix) 
fit2 <- eBayes(fit2)
deg_df <- topTable(fit2, coef=1, adjust="fdr", sort.by="B", number=Inf)
deg_df <- deg_df[!is.na(deg_df$adj.P.Val),] # Removes rows with NA values
head(deg_df)[,22:27]
```

# Enrichment analysis of GO terms 

The following performs over-represention analysis (ORA) of GO terms using functions from 
the `clusterProfiler` package [@Yu2012-hr].

## Prepare input

```{r overrepresentation_analysis, eval=TRUE, message=FALSE}
cutoff <- 0.05 # Cutoff to use for filtering on adjusted p-value (FDR)
ids <- deg_df[deg_df$adj.P.Val <= cutoff, "Gene.ID"]
ids <- ids[!grepl("/", ids)] # Removes ambiguous mappings
ids <- ids[grepl("", ids)]
```

## Enrichment of MF terms

```{r overrepresentation_analysis_mf, eval=TRUE, message=FALSE}
library(clusterProfiler); library(org.Hs.eg.db); library(enrichplot)
ego_mf <- enrichGO(gene=ids, OrgDb=org.Hs.eg.db, ont="MF", pAdjustMethod="BH", pvalueCutoff=0.1, readable=TRUE)
dim(ego_mf)
head(ego_mf)
```

## Visualization of MF result

```{r plot_enrichment_mf, eval=TRUE, message=FALSE}
barplot(ego_mf, showCategory=10)
goplot(ego_mf)
```

## Enrichment of BP terms

```{r overrepresentation_analysis_bp, eval=TRUE, message=FALSE}
ego_bp <- enrichGO(gene=ids, OrgDb=org.Hs.eg.db, ont="BP", pAdjustMethod="BH", pvalueCutoff=0.1, readable=TRUE)
dim(ego_bp)
head(ego_bp)
```

## Visualization of BP result
```{r plot_enrichment_bp, eval=TRUE, message=FALSE}
barplot(ego_bp, showCategory=10)
```

# Clustering

...

```{r clustering, eval=TRUE, message=FALSE}
library(pheatmap); library("RColorBrewer")
cutoff <- 0.05 # Cutoff to use for filtering on adjusted p-value (FDR)
affy_ids <- row.names(deg_df[deg_df$adj.P.Val <= cutoff, ])
deg_ma <- exprs(eset)[affy_ids, ] 
pheatmap(deg_ma, scale="row", color=brewer.pal(9, "Blues"))
```

# Session Info

```{r sessionInfo}
sessionInfo()
```

# References
